Homework-4.py

```diff
@@ -0,0 +1,261 @@
+
+    group1 = df[df[response] == 0]
+    group2 = df[df[response] == 1]
+
+    hist_data = [group1[predictor], group2[predictor]]
+
+    # Create distribution plot with custom bin_size
+    fig_1 = ff.create_distplot(hist_data, group_labels, bin_size=10)
+    fig_1.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title="Predictor",
+        yaxis_title="Distribution",
+    )
+    fig_1.show()
+
+    count, bins_edges = np.histogram(df[predictor], bins=4)
+
+    # print(count)
+    # print(bins_edges)
+    bins_weighted_mean_diff = [0 for i in range(len(bins_edges))]
+    bins_unweighted_mean_diff = [0 for i in range(len(bins_edges))]
+    print()
+
+    pop_mean = np.mean(df[predictor])
+    print("population mean for predictor ", predictor, " = ", pop_mean)
+
+    for i in range(len(bins_edges) - 1):
+        bin_data = df[
+            (bins_edges[i + 1] >= df[predictor]) & (bins_edges[i] < df[predictor])
+            ]
+# step 5.Difference with mean of response along with its plot (weighted and unweighted)
+        # print(bin_data[predictor])
+        # print()
+        # weighted mean or response
+        bins_weighted_mean_diff[i] = np.square(
+            count[i] * (np.mean(bin_data[predictor]) - pop_mean)
+        )
+        # unweighted mean or response
+        bins_unweighted_mean_diff[i] = np.square(
+            (np.mean(bin_data[predictor]) - pop_mean)
+        )
+        print(
+            "for predictor ",
+            predictor,
+            "squared diff of mean for bin ",
+            i,
+            " is ",
+            bins_weighted_mean_diff[i],
+        )
+
+    print("w mean squared difference = ", np.mean(bins_weighted_mean_diff))
+    stats_table["w_mean_diff"][predictor] = np.mean(bins_weighted_mean_diff)
+    stats_table["uw_mean_diff"][predictor] = np.mean(bins_unweighted_mean_diff)
+
+
+# if response variable is continous and predictor is continous.
+def plot_cont_response_cont_predictor(df, predictor, response):
+    n = 200
+    x = df[predictor]
+    y = df[response]
+
+    fig = px.scatter(x=x, y=y, trendline="ols")
+    fig.update_layout(
+        title="Continuous Response by Continuous Predictor",
+        xaxis_title="Predictor",
+        yaxis_title="Response",
+    )
+    fig.show()
+
+    return
+
+# step 6. p-value & t-score (continuous predictors only) along with it's plot
+# logistic regression will work if response variable it categorical
+def logistic_regression(df, predictor_name, response, stats_table):
+    x = df[predictor_name]
+    y = df[response]
+
+    # Remaking df with just these two columns to remove na's
+    df = pd.DataFrame({predictor_name: x, response: y})
+    pd.set_option("mode.use_inf_as_na", True)
+    df = df.dropna()
+    x = df[predictor_name]
+    y = df[response].map(int)
+
+    predictor = statsmodels.api.add_constant(x)
+    logistic_regression_model = statsmodels.api.Logit(np.asarray(y), np.asarray(predictor))
```

Comment:
Please bear in mind that plotly does not support logistic regression trendlines. So the plots will reflect linear probability models, not logit regressions.